File: LLD-Python/solid_principle/dependency_inversion_principle/dependency_inversion_good.py
from abc import ABC, abstractmethod

class MessageSender(ABC):
    @abstractmethod
    def send(self, message: str) -> None:
        ...

class EmailSender(MessageSender):
    def send(self, message: str) -> None:
        # send email
        pass

class NotificationService:
    def __init__(self, sender: MessageSender) -> None:
        self.sender = sender

    def notify(self, message: str)-> None:
        self.sender.send(message)

File: LLD - Java/src/main/java/org/solid/dependencyInversionPrinciple/DependencyInversionGood.java
package org.solid.dependencyInversionPrinciple;

public class DependencyInversionGood {
}

interface MessageSender {
    void send(String message);
}

class EmailSender implements MessageSender {
    public void send(String message) { /* send email */}
}

class NotificationService {
    private final MessageSender sender;

    NotificationService(MessageSender sender){
        this.sender = sender;
    }

    void notify(String message){
        sender.send(message);
    }
}

